sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,888 @@
+
+        .. deprecated:: 0.20
+            The `categorical_features` keyword is deprecated and will be
+            removed in 0.22. You can use the ``ColumnTransformer`` instead.
+
+    Attributes
+    ----------
+    categories_ : list of arrays
+        The categories of each feature determined during fitting
+        (in order corresponding with output of ``transform``).
+
+    active_features_ : array
+        Indices for active features, meaning values that actually occur
+        in the training set. Only available when n_values is ``'auto'``.
+
+        .. deprecated:: 0.20
+
+    feature_indices_ : array of shape (n_features,)
+        Indices to feature ranges.
+        Feature ``i`` in the original data is mapped to features
+        from ``feature_indices_[i]`` to ``feature_indices_[i+1]``
+        (and then potentially masked by `active_features_` afterwards)
+
+        .. deprecated:: 0.20
+
+    n_values_ : array of shape (n_features,)
+        Maximum number of values per feature.
+
+        .. deprecated:: 0.20
+
+    Examples
+    --------
+    Given a dataset with two features, we let the encoder find the unique
+    values per feature and transform the data to a binary one-hot encoding.
+
+    >>> from sklearn.preprocessing import OneHotEncoder
+    >>> enc = OneHotEncoder(handle_unknown='ignore')
+    >>> X = [['Male', 1], ['Female', 3], ['Female', 2]]
+    >>> enc.fit(X)
+    ... # doctest: +ELLIPSIS
+    OneHotEncoder(categories='auto', dtype=<... 'numpy.float64'>,
+           handle_unknown='ignore', sparse=True)
+
+    >>> enc.categories_
+    [array(['Female', 'Male'], dtype=object), array([1, 2, 3], dtype=object)]
+    >>> enc.transform([['Female', 1], ['Male', 4]]).toarray()
+    array([[ 1., 0., 1., 0., 0.],
+           [ 0., 1., 0., 0., 0.]])
+    >>> enc.inverse_transform([[0, 1, 1, 0, 0], [0, 0, 0, 1, 0]])
+    array([['Male', 1],
+           [None, 2]], dtype=object)
+
+    See also
+    --------
+    sklearn.preprocessing.OrdinalEncoder : performs an ordinal (integer)
+      encoding of the categorical features.
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
+    sklearn.preprocessing.LabelBinarizer : binarizes labels in a one-vs-all
+      fashion.
+    sklearn.preprocessing.MultiLabelBinarizer : transforms between iterable of
+      iterables and a multilabel format, e.g. a (samples x classes) binary
+      matrix indicating the presence of a class label.
+    """
+
+    def __init__(self, n_values=None, categorical_features=None,
+                 categories=None, sparse=True, dtype=np.float64,
+                 handle_unknown='error'):
+        self._categories = categories
+        if categories is None:
+            self.categories = 'auto'
```

Comment:
Opened #11117 I think this is not good.

Reply:
Hmm, do you have an alternative suggestion how to deal with this? 

I basically wanted to ensure that the `.categories` attribute (and the value you see in `get_params` or in the repr ) embodies the *actual* default ('auto'), and not the `None` whose only purpose is to check if the user passed the value for `categories` explicitly or not.

Reply:
Again, this is not how we have been doing things. It would be nice to have the "right" value in the ``__repr__`` and ``get_params``, but do we want to start this now? @GaelVaroquaux @jnothman  opinions? So far we did everything in ``fit`` and basically didn't care about repr and get_params being weird for deprecations.

Reply:
The main reason I find it important in this case is that I want to make the experience for *new* users of the OneHotEncoder (the ones that don't have the baggage of how it was before, and just want to start using the new functionality of finally being able to one-hot-encode string categories) as smooth as possible. And those don't need to wonder what this None means (although they will see it in the signature of course), and they don't need to see the deprecated keywords in the repr.

To be clear, I don't have problems with being pointed out how things are typically done in sklearn (I don't yet have much experience on that with the rest of the code base), so happy to change things to be more in line with sklearn development practices. 
I just wanted to try to give a best possible user experience, but sorry if that raised different discussions on changing standard practices :-)

Reply:
I think that's a good goal to have in mind :) And I'm not saying "no", I'm just saying if we want to change our practices we should give it due consideration.

Reply:
BTW, the reason the common tests were not failing on this, is because they are skipped for OneHotEncoder ... (https://github.com/scikit-learn/scikit-learn/blob/4b24fbedf5fa3b7b6b559141ad78708145b09704/sklearn/utils/testing.py#L553)